### DISCUSSION TITLE
Issue with my RLS

### DISCUSSION THREAD

**Discussion by ricardosikic** [supabase/supabase#11847] [Questions] [answered]
Hello,
I currently have an RLS to do `update` on the `projects` table where `uid() = user_id`. The problem arose for me now because when a person contributes to a project a record of a transaction is created which, with a script, updates the amount contributed to the project in the `projects` table, but it doesn't let me update (unless I remove the RLS from that table) as the logged in contributing user is not the same as the one in the `projects.user_id` column.
So it occurs to me that I would have to add to the RLS of that `projects` table a permission that says `money_transactions.donor_user_id = uid()` or just take out the rls and handle these `delete permissions with validations in the `front` `and `update`?

**Comment by GaryAustin1** [comment, marked as answer]
Not a whole lot of info to go on.
What is a "script"? 
Is the owner of a project the only one allowed to update projects or any signed in user?  If so then just using authenticated user might work.
Is there a list of ids that "belong" to a project and they all have access? 

You can use an rpc function that would allow you to update the column and have it ignore RLS and just check for what ever security you need in it as one possibility.
You could also split the contribution out to a separate table so it can have it's own privileges and just use a foreign key back to the project.

**Reply by ricardosikic** [reply]
Hello, thanks for the response @GaryAustin1 . But i'm having some issues trying to update the record with rpc. i have created a function in `sql editor` to update the column but is not updating the record.
```
create function update_initiative_amount(amount integer, initiative_id integer)
returns TABLE(id integer) as $$
  UPDATE initiatives
  SET amount_collected = amount
  WHERE id = initiative_id
  RETURNING id
$$ language sql;
```

The front is calling the rpc function:
```
  const updateInitiativeAmount = async(initiativeCreatorId, initiativeId, amount) => {

    try {
      const { data } = await supabase.rpc('update_initiative_amount', {amount: amount, initiative_id: initiativeId});
      console.log('result ', data);
    } catch (err) {
      console.log('error ', err);
    }
  }
```

**Reply by GaryAustin1** [reply]
You don't have "security definer" type in the function so it still has to meet RLS if that applies to the user calling it.  BUT if you make it security definer then you need to make sure to check for the user in the function is allowed to do the operation.

**Reply by ricardosikic** [reply]
with the RLS only the user authenticated = initiative.user_id is allowed to update. That is the main reason for not being able to update after a connected user other than the project user contributes to it so that the record is updated with the new amount. It would only let me do it if I were the same owner of the project that is connected in the session and that contributed to the project.

**Reply by GaryAustin1** [reply]
Yes, this a a reason to use “security definer” to allow operations to bypass the standard RLS. Just you then need to make sure you don’t allow an anon user to execute the function with code in it.

**Reply by ricardosikic** [reply]
@GaryAustin1 you are the best!. `SECURITY DEFINER` was missing.
```
create function update_initiative_amount_one(amount integer, initiative_id integer)
returns 
TABLE(id integer)
SECURITY DEFINER 
as $$
  UPDATE initiatives
  SET amount_collected = amount
  WHERE id = initiative_id
  RETURNING id
$$ language sql;
```

**Reply by GaryAustin1** [reply]
That is good, but before you do this in production, please learn about security definer functions (setting search path is usually recommended) and right now any user in with access to your anon key can change that amount as you are not even checking for current_user = authenticated at least.

**Reply by ricardosikic** [reply]
ok i will learn about it

**Reply by ricardosikic** [reply]
@GaryAustin1 You mean the anon_key that I have set in the `.env` file

**Reply by GaryAustin1** [reply]
Probably, but I refer to the publicly visible Supabase anon_key use in all your calls from a browser.
I'm simplifying below.
Without RLS anyone can access your tables using your client code or not.  Any code running on the browser is not secure, even for a signed in user.  You have to have RLS.

The security definer RPC function is sometimes needed for special situations where you need to bypass RLS, but it does so for anyone with access to your database signed in or not, it runs like user Postgres by default, which bypasses RLS.  So you have to make sure, if you don't want anyone to use the code in that function to modify tables, that you check for them in the WHERE statement if SQL function, or in if statements in a plpgsql function.

You might also consider Supabase's Discord channel for questions on how to do things as it is more informal and interactive.
